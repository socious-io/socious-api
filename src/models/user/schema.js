import Joi from 'joi';
import {SocialCauses} from '../../utils/types.js';

export const updateProfileSchem = Joi.object({
  first_name: Joi.string().required(),
  last_name: Joi.string().required(),
  bio: Joi.string(),
  city: Joi.string(),
  address: Joi.string(),
  wallet_address: Joi.string(),
  avatar: Joi.string().uuid(),
  cover_image: Joi.string().uuid(),
  social_causes: Joi.array().items(
    Joi.string().valid(...Object.values(SocialCauses)),
  ),
  skills: Joi.array().items(Joi.string())
});
