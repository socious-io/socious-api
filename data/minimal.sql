CREATE TABLE public.users(
  id serial PRIMARY KEY,
  first_name varchar(70) DEFAULT NULL,
  username varchar(200) UNIQUE NOT NULL,
  name varchar(255) DEFAULT NULL,
  email varchar(200) UNIQUE NOT NULL,
  email_text varchar(255) DEFAULT NULL,
  mobile_countries_id int(11) DEFAULT NULL,
  phone varchar(255) DEFAULT NULL,
  wallet_address varchar(255) DEFAULT NULL,
  password text NOT NULL,
  remember_token text DEFAULT NULL,
  city text,
  city_ja mediumtext,
  description_search mediumtext,
  description_search_ja mediumtext,
  address_detail text,
  avatar text,
  cover_image text,
  expiry_date datetime DEFAULT NULL,
  active boo,
  activation_token varchar(255) DEFAULT NULL,
  mission text,
  bio mediumtext,
  profile_id varchar(200) DEFAULT NULL,
  view_as int(11) DEFAULT NULL,
  language varchar(255) DEFAULT NULL,
  my_conversation varchar(255) DEFAULT NULL,  
  email_verified_at datetime DEFAULT NULL,
  impact_score float NOT NULL DEFAULT '0',
  lock_otp datetime DEFAULT NULL,
  created_at datetime NOT NULL DEFAULT CURRENT_TIMESTAMP(6),
  updated_at datetime NOT NULL DEFAULT CURRENT_TIMESTAMP(6) ON UPDATE CURRENT_TIMESTAMP(6),
  deleted_at datetime DEFAULT NULL
);
