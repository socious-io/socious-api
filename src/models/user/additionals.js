import sql from 'sql-template-tag'
import { app } from '../../index.js'
import { EntryError } from '../../utils/errors.js'

export const addLanguage = async (user, { name, level }) => {
  try {
    const { rows } = await app.db.query(sql`
    INSERT INTO languages (name, level, user_id) 
    VALUES (${name}, ${level}, ${user.id})
    RETURNING *
  `)
    return rows[0]
  } catch (err) {
    throw new EntryError(err.message)
  }
}

export const editLanguage = async (id, user, { name, level }) => {
  try {
    const { rows } = await app.db.query(sql`
    UPDATE languages SET 
    name=${name},
    level=${level}
    WHERE user_id=${user.id} AND id=${id}
    RETURNING *
  `)
    return rows[0]
  } catch (err) {
    throw new EntryError(err.message)
  }
}

export const removeLanguage = async (id, user) => {
  await app.db.query(sql`DELETE FROM languages WHERE id=${id} AND user_id=${user.id}`)
}

export const addExperience = async (
  user,
  { org_id, title, description, skills, start_at, end_at, city, country, employment_type, job_category_id }
) => {
  try {
    const { rows } = await app.db.query(sql`
    INSERT INTO experiences (
      org_id,
      title,
      description,
      skills,
      start_at,
      end_at,
      user_id,
      city,
      country,
      employment_type,
      job_category_id
    ) 
    VALUES (
      ${org_id},
      ${title},
      ${description},
      ${skills},
      ${start_at},
      ${end_at},
      ${user.id},
      ${city},
      ${country},
      ${employment_type},
      ${job_category_id}
      )
    RETURNING *
  `)
    return rows[0]
  } catch (err) {
    throw new EntryError(err.message)
  }
}

export const editExperience = async (
  id,
  user,
  { org_id, title, description, skills, start_at, end_at, city, country, employment_type, job_category_id }
) => {
  try {
    const { rows } = await app.db.query(sql`
    UPDATE experiences SET
      org_id=${org_id},
      title=${title},
      description=${description},
      skills=${skills},
      start_at=${start_at},
      end_at=${end_at},
      city=${city},
      country=${country},
      employment_type=${employment_type},
      job_category_id=${job_category_id}
    WHERE id=${id} AND user_id=${user.id}
    RETURNING *
  `)
    return rows[0]
  } catch (err) {
    throw new EntryError(err.message)
  }
}

export const removeExperience = async (id, user) => {
  await app.db.query(sql`DELETE FROM experiences WHERE id=${id} AND user_id=${user.id}`)
}
