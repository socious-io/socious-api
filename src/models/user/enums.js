export const StatusType = {
  ACTIVE: 'ACTIVE',
  INACTIVE: 'INACTIVE',
  SUSPEND: 'SUSPEND'
}
