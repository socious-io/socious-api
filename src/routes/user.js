import Router from '@koa/router'
import User from '../models/user/index.js'
import Applicant from '../models/applicant/index.js'
import Auth from '../services/auth/index.js'
import Mission from '../models/mission/index.js'
import Offer from '../models/offer/index.js'
import Skill from '../models/skill/index.js'
import ImpactPoints from '../services/impact_points/index.js'
import Payment from '../services/payments/index.js'
import { paginate } from '../utils/middlewares/requests.js'
import { loginOptional, loginRequired } from '../utils/middlewares/authorization.js'
import Data, { validate } from '@socious/data'
import { checkIdParams } from '../utils/middlewares/route.js'
import { putContact } from '../services/sendgrid/index.js'
import { BadRequestError } from '../utils/errors.js'
import { recommendUserByUser, recommendProjectByUser, recommendOrgByUser } from '../services/recommender/index.js'

export const router = new Router()

router.get('/:id/profile', loginOptional, checkIdParams, async (ctx) => {
  if (ctx.user.status === User.StatusType.INACTIVE) {
    ctx.body = await User.getProfileLimited(ctx.params.id)
    return
  }
  ctx.body = await User.getProfile(ctx.params.id, ctx.identity.id)
})

router.post('/:id/report', loginRequired, async (ctx) => {
  await validate.ReportSchema.validateAsync(ctx.request.body)
  await User.report({
    ...ctx.request.body,
    user_id: ctx.params.id,
    identity_id: ctx.identity.id
  })

  ctx.body = {
    message: 'success'
  }
})

router.get('/by-username/:username/profile', loginOptional, async (ctx) => {
  if (ctx.user.status === User.StatusType.INACTIVE) {
    ctx.body = await User.getProfileByUsernameLimited(ctx.params.username)
    return
  }
  ctx.body = await User.getProfileByUsername(ctx.params.username, ctx.identity.id)
})

router.get('/profile', loginRequired, async (ctx) => {
  ctx.body = await User.getProfile(ctx.user.id)
})

router.post('/update/wallet', loginRequired, async (ctx) => {
  const wallet = ctx.request.body.wallet_address
  if (!wallet) throw new BadRequestError('wallet address is required')
  await User.updateWalletAddress(ctx.user.id, wallet)
  ctx.body = await User.getProfile(ctx.user.id)
})

router.post('/update/profile', loginRequired, async (ctx) => {
  await validate.UpdateProfileSchema.validateAsync(ctx.request.body)

  const skills = await Skill.getAllByNames(ctx.request.body.skills)
  ctx.request.body.skills = skills.map((s) => s.name)

  ctx.body = await User.updateProfile(ctx.user.id, ctx.request.body)

  putContact({
    first_name: ctx.body.first_name,
    last_name: ctx.body.last_name,
    email: ctx.body.email
  })
})

router.post('/change-password', loginRequired, async (ctx) => {
  await Auth.changePassword(ctx.user, ctx.request.body)
  ctx.body = {
    message: 'success'
  }
})

router.post('/change-password-direct', loginRequired, async (ctx) => {
  await Auth.directChangePassword(ctx.user, ctx.request.body)

  ctx.body = {
    message: 'success'
  }
})

router.post('/delete', loginRequired, async (ctx) => {
  await User.remove(ctx.user, ctx.request.body.reason)
  ctx.body = {
    message: 'success'
  }
})

router.get('/applicants', loginRequired, checkIdParams, paginate, async (ctx) => {
  ctx.body = await Applicant.getByUserId(ctx.user.id, ctx.paginate)
})

router.get('/:id/missions', loginRequired, checkIdParams, paginate, async (ctx) => {
  ctx.paginate.filter.assignee_id = ctx.params.id
  ctx.body = await Mission.getAll(ctx.paginate)
})

router.get('/missions', loginRequired, paginate, async (ctx) => {
  ctx.paginate.filter.assignee_id = ctx.identity.id

  const missions = await Mission.getAll(ctx.paginate)

  ctx.body = missions.map((m) => {
    return {
      ...m,
      ...Payment.amounts({
        amount: m.offer.assignment_total,
        service: Data.PaymentService.STRIPE
          ? m.offer.payment_mode === Data.PaymentMode.FIAT
          : Data.PaymentService.CRYPTO,
        verified: m.assigner.meta.verified_impact
      })
    }
  })
})

router.get('/offers', loginRequired, paginate, async (ctx) => {
  ctx.body = await Offer.getAll(ctx.identity.id, ctx.paginate)
})

router.post('/languages', loginRequired, async (ctx) => {
  await validate.ProfileAddLanguageSchema.validateAsync(ctx.request.body)
  ctx.body = await User.addLanguage(ctx.user, ctx.request.body)
})

router.post('/languages/update/:id', loginRequired, checkIdParams, async (ctx) => {
  await validate.ProfileAddLanguageSchema.validateAsync(ctx.request.body)
  ctx.body = await User.editLanguage(ctx.params.id, ctx.user, ctx.request.body)
})

router.post('/languages/remove/:id', loginRequired, checkIdParams, async (ctx) => {
  await User.removeLanguage(ctx.params.id, ctx.user)
  ctx.body = { message: 'success' }
})

router.post('/experiences', loginRequired, async (ctx) => {
  await validate.ProfileExperienceSchema.validateAsync(ctx.request.body)
  ctx.body = await User.addExperience(ctx.user, ctx.request.body)
})

router.post('/experiences/update/:id', loginRequired, checkIdParams, async (ctx) => {
  await validate.ProfileExperienceSchema.validateAsync(ctx.request.body)
  ctx.body = await User.editExperience(ctx.params.id, ctx.user, ctx.request.body)
})

router.post('/experiences/remove/:id', loginRequired, checkIdParams, async (ctx) => {
  ctx.body = await User.removeExperience(ctx.params.id, ctx.user)
})

router.get('/badges', loginRequired, async (ctx) => {
  ctx.body = { badges: await ImpactPoints.badges(ctx.identity.id) }
})

router.get('/:id/badges', loginRequired, checkIdParams, async (ctx) => {
  ctx.body = { badges: await ImpactPoints.badges(ctx.params.id) }
})

router.get('/impact-points', loginRequired, paginate, async (ctx) => {
  ctx.body = await ImpactPoints.history(ctx.identity.id, ctx.paginate)
})

router.post('/open-to-work', loginRequired, async (ctx) => {
  ctx.body = {
    open_to_work: await User.openToWork(ctx.user.id)
  }
})

router.post('/open-to-volunteer', loginRequired, async (ctx) => {
  ctx.body = {
    open_to_volunteer: await User.openToVolunteer(ctx.user.id)
  }
})

router.get('/:username/recommend/jobs', loginOptional, async (ctx) => {
  ctx.body = await recommendProjectByUser(ctx.params.username)
})

router.get('/:username/recommend/users', loginOptional, async (ctx) => {
  ctx.body = await recommendUserByUser(ctx.params.username)
})

router.get('/:username/recommend/orgs', loginOptional, async (ctx) => {
  ctx.body = await recommendOrgByUser(ctx.params.username)
})
