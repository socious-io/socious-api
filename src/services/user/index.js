import * as write from './write.js';
import * as read from './read.js';
import * as auth from './auth.js';

export default {
  ...auth,
  ...read,
  ...write,
};
