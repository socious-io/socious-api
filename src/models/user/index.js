import * as write from './write.js'
import * as read from './read.js'
import * as enums from './enums.js'
import * as additionals from './additionals.js'

export default {
  ...enums,
  ...read,
  ...write,
  ...additionals
}
