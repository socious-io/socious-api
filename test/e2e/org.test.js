import supertest from 'supertest'
import { app } from '../../src/index.js'
import data from '../data/index.js'
import { registerAndVerify } from './globals/user.js'
import { create, getAll, getFiltered } from './globals/org.js'

let server, request

beforeAll(async () => {
  server = app.listen()
  request = supertest(server)

  await registerAndVerify(request, data)
})

test('create', async () => create(request, data))
test('get all', async () => getAll(request, data))
test('get filtered', async () => getFiltered(request, data))

const cleanup = async () => {
  await app.db.query('DELETE FROM users')
  await app.db.query('DELETE FROM organizations')
  await app.db.pool.end()
}

afterAll((done) => {
  cleanup(done).then(() => {
    server.close(done)
  })
})
