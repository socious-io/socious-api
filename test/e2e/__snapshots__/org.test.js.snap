// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
{
  "error": ""email" must be a valid email",
}
`;

exports[`create 1`] = `
{
  "address": "address",
  "bio": "test",
  "city": "Tokyo",
  "country": "JP",
  "cover_image": null,
  "created_at": Any<String>,
  "created_by": Any<String>,
  "culture": "test",
  "description": "test",
  "email": "test@socious.io",
  "followers": 0,
  "followings": 0,
  "geoname_id": null,
  "hiring": false,
  "id": Any<String>,
  "image": null,
  "impact_points": 0,
  "industry": null,
  "mission": "test",
  "mobile_country_code": null,
  "name": "test ORG",
  "other_party_id": null,
  "other_party_title": null,
  "other_party_url": null,
  "phone": "9379911",
  "search_tsv": Any<String>,
  "shortname": Any<String>,
  "size": null,
  "social_causes": [
    "SOCIAL",
    "HEALTH",
  ],
  "status": "ACTIVE",
  "type": "COOP",
  "updated_at": Any<String>,
  "verified_impact": false,
  "wallet_address": null,
  "website": "http://test.com",
}
`;

exports[`create 2`] = `
{
  "address": "address",
  "bio": "test",
  "city": "Tehran",
  "country": "IR",
  "cover_image": null,
  "created_at": Any<String>,
  "created_by": Any<String>,
  "culture": "test",
  "description": "test",
  "email": "test@socious.io",
  "followers": 0,
  "followings": 0,
  "geoname_id": null,
  "hiring": false,
  "id": Any<String>,
  "image": null,
  "impact_points": 0,
  "industry": null,
  "mission": "test",
  "mobile_country_code": null,
  "name": "test ORG",
  "other_party_id": null,
  "other_party_title": null,
  "other_party_url": null,
  "phone": "9899993",
  "search_tsv": Any<String>,
  "shortname": Any<String>,
  "size": null,
  "social_causes": [
    "SOCIAL",
    "MENTAL",
  ],
  "status": "ACTIVE",
  "type": "SOCIAL",
  "updated_at": Any<String>,
  "verified_impact": false,
  "wallet_address": null,
  "website": "http://test.com",
}
`;

exports[`create 3`] = `
{
  "error": ""phone" must be a string",
}
`;

exports[`get all 1`] = `
{
  "items": [
    {
      "address": "address",
      "benefits": null,
      "bio": "test",
      "city": "Tehran",
      "country": "IR",
      "cover_image": null,
      "created_at": Any<String>,
      "created_by": Any<String>,
      "culture": "test",
      "description": "test",
      "email": "test@socious.io",
      "followers": 1,
      "followings": 0,
      "geoname_id": null,
      "hiring": false,
      "id": Any<String>,
      "image": null,
      "impact_points": 0,
      "industry": null,
      "mission": "test",
      "mobile_country_code": null,
      "name": "test ORG",
      "other_party_id": null,
      "other_party_title": null,
      "other_party_url": null,
      "phone": "9899993",
      "recommendations": null,
      "search_tsv": Any<String>,
      "shortname": Any<String>,
      "size": null,
      "social_causes": [
        "SOCIAL",
        "MENTAL",
      ],
      "status": "ACTIVE",
      "type": "SOCIAL",
      "updated_at": Any<String>,
      "verified_impact": false,
      "wallet_address": null,
      "website": "http://test.com",
    },
    {
      "address": "address",
      "benefits": null,
      "bio": "test",
      "city": "Tokyo",
      "country": "JP",
      "cover_image": null,
      "created_at": Any<String>,
      "created_by": Any<String>,
      "culture": "test",
      "description": "test",
      "email": "test@socious.io",
      "followers": 1,
      "followings": 0,
      "geoname_id": null,
      "hiring": false,
      "id": Any<String>,
      "image": null,
      "impact_points": 0,
      "industry": null,
      "mission": "test",
      "mobile_country_code": null,
      "name": "test ORG",
      "other_party_id": null,
      "other_party_title": null,
      "other_party_url": null,
      "phone": "9379911",
      "recommendations": null,
      "search_tsv": Any<String>,
      "shortname": Any<String>,
      "size": null,
      "social_causes": [
        "SOCIAL",
        "HEALTH",
      ],
      "status": "ACTIVE",
      "type": "COOP",
      "updated_at": Any<String>,
      "verified_impact": false,
      "wallet_address": null,
      "website": "http://test.com",
    },
  ],
  "limit": 10,
  "page": 1,
  "total_count": 2,
}
`;

exports[`get filtered 1`] = `
{
  "items": [
    {
      "address": "address",
      "benefits": null,
      "bio": "test",
      "city": "Tokyo",
      "country": "JP",
      "cover_image": null,
      "created_at": Any<String>,
      "created_by": Any<String>,
      "culture": "test",
      "description": "test",
      "email": "test@socious.io",
      "followers": 1,
      "followings": 0,
      "geoname_id": null,
      "hiring": false,
      "id": Any<String>,
      "image": null,
      "impact_points": 0,
      "industry": null,
      "mission": "test",
      "mobile_country_code": null,
      "name": "test ORG",
      "other_party_id": null,
      "other_party_title": null,
      "other_party_url": null,
      "phone": "9379911",
      "recommendations": null,
      "search_tsv": Any<String>,
      "shortname": Any<String>,
      "size": null,
      "social_causes": [
        "SOCIAL",
        "HEALTH",
      ],
      "status": "ACTIVE",
      "type": "COOP",
      "updated_at": Any<String>,
      "verified_impact": false,
      "wallet_address": null,
      "website": "http://test.com",
    },
  ],
  "limit": 10,
  "page": 1,
  "total_count": 1,
}
`;
